presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorBucketFunction.java

```diff
@@ -15,34 +15,91 @@
 
 import com.facebook.presto.spi.BucketFunction;
 import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PrestoException;
 import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.annotations.VisibleForTesting;
 import io.airlift.slice.XxHash64;
 
+import java.util.List;
+
+import static com.facebook.presto.spi.StandardErrorCode.NOT_SUPPORTED;
 import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.Varchars.isVarcharType;
 import static com.google.common.base.Preconditions.checkArgument;
 
 public class RaptorBucketFunction
         implements BucketFunction
 {
-    private final int bucketCount;
+    private final HashFunction[] functions;
 
-    public RaptorBucketFunction(int bucketCount)
+    public RaptorBucketFunction(List<Type> types)
     {
-        checkArgument(bucketCount > 0, "bucketCount must be at least one");
-        this.bucketCount = bucketCount;
+        checkArgument(types != null, "types is is null");
+        checkArgument(!types.isEmpty(), "types is empty");
+
+        this.functions = new HashFunction[types.size()];
+        for (int i = 0; i < types.size(); i++) {
+            Type type = types.get(i);
+            if (type.equals(BIGINT)) {
```

Comment:
Let's extract a helper function that goes from `Type` to `HashFunction`, then use that to implement `checkTypeSupported`, so that we don't have to repeat the list of supported types.

It's important that they don't get out of sync, otherwise you could create a table that can't be used.